sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -601,8 +602,17 @@ public static Read read() {
 
     @Override
     public TableRow apply(SchemaAndRecord schemaAndRecord) {
-      return BigQueryAvroUtils.convertGenericRecordToTableRow(
-          schemaAndRecord.getRecord(), schemaAndRecord.getTableSchema());
+      // TODO(BEAM-9114): Implement a function to encapsulate row conversion logic.
+      try {
+        return BigQueryAvroUtils.convertGenericRecordToTableRow(
+            schemaAndRecord.getRecord(), schemaAndRecord.getTableSchema());
+      } catch (IllegalStateException i) {
+        if (schemaAndRecord.getRow() != null) {
+          return BigQueryUtils.toTableRow().apply(schemaAndRecord.getRow());
+        }
+        throw new IllegalStateException(
+            "Record should be of instance GenericRecord (for Avro format) or of instance Row (for Arrow format), but it is not.");
+      }
```

Comment:
This feels pretty hacky. Ideally I think we would use `Row` as an intermediate format that can be backed by either an Avro `GenericRecord` or an Arrow `RecordBatch`. But if that's too tricky to get right we should at least avoid branching within a `DoFn` like this, since it will mean checking the type of every single element. We could instead produce different logic at pipeline construction time.

Reply:
So, the idea would be to have this apply function to only use `Row` from `SchemaAndRecord,` and have a function that encapsulates converting `GenericRecord` and `RecordBatch` to `Row` to use that before this apply?

Reply:
Yes I think that would be the ideal solution. Note there's some logic for mapping Avro `GenericRecord` to Beam `Row` already in [`AvroUtils`](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java)

Reply:
Hi, I made the other changes but I'm struggling with this one, so Arrow `RecordBatch` should be mapped to a single `Row`? or is it `ArrowRecordBatch`? I'm a bit confused about the right data type. Also, is there any example on how to map the data, I saw that there is a `VectorSchemaRoot` to `Iterable<Row>` conversion already that uses a Factory of ListValueGetters to do the mapping, would something similar be needed?

Reply:
@TheNeuralBit Hi, could you help point me in the right direction with this questions?

Reply:
Hi @MiguelAnzoWizeline - sorry I missed this. An Arrow `RecordBatch` contains multiple records, and each needs to map to its own `Row` instance. I don't think you actually need to worry about that so much to do what I'm proposing, the code you mentioned to convert a `VectorSchemaRoot` to `Iterable<Row>` is already handling this. That's why for the ARROW path, `SchemaAndRecord` contains a `Row` instance.

The problem is that for the AVRO path, `SchemaAndRecord` contains an Avro `GenericRecord`, which is why we need to have this kind of hacky check here to handle either `Row` or `GenericRecord`. What I'm suggesting is that we should use the more generic `Row` for this path too, since there's code for making a `Row` out of a `GenericRecord` in `AvroUtils`.

Reply:
Does that make sense?

Reply:
Hi, yes, that makes more sense, thanks, I will be making the change and updating.

Reply:
R: @TheNeuralBit